File: src/instruction.rs
use borsh::{BorshDeserialize, BorshSerialize};
use solana_program::borsh::try_from_slice_unchecked;

#[derive(BorshDeserialize, BorshSerialize)]
pub enum InstructionEnum {
    InitProgram,
    RegisterUser{user_name: Vec<u8>},
    RewardXP{total_xp: u64, user_name: Vec<u8>},
    BlankInstruction,
}

impl InstructionEnum {
    pub fn decode(data: &[u8]) -> Self {
        try_from_slice_unchecked(data).unwrap()
    }
}

File: python-client/instruction.py
from borsh_construct import *
InstructionEnum = Enum(
    "InitProgram",
    "RegisterUser" / CStruct("user_name" / Vec(U8)),
    "RewardXP" / CStruct("total_xp" / U64, "user_name" / Vec(U8)),
    "BlankInstruction",

    enum_name = "InstructionEnum"
)

def build_instruction(instruction):
    return InstructionEnum.build(instruction)

    